### ISSUE TITLE
High load in 1) SELECT queries when we have 7000 rows and 21000 branches 2) create extra branches when we've already have 21000

### ISSUE THREAD

**Issue by nkonev** [dolthub/dolt#5175]
I saw very similar https://github.com/dolthub/dolt/issues/5148 but I'm not sure if there is the same isuue.

So this test can be synthetic, but I think it worth to share the results.

Version is 0.52.8

# Problem 1 - high CPU when I use sort

Reproduction:
```
git clone git@github.com:nkonev/companies.git
cd companies
git checkout select-high-cpu 
docker-compose up -d
```

### Run web app in the first terminal window
```
./mvnw spring-boot:run
```

### Then create 7000 companies in the another terminal
```
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#seven_thousand_companies" test
```

### Then create branch for every company and add a commit three times (sequentially)
```
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
```

### Then run every-second requests in three terminal windows in parallel
```
watch -n 1 curl -Ss --url 'http://localhost:8080/company'
```

```
watch -n 1 curl -Ss --url 'http://localhost:8080/company'
```

```
watch -n 1 curl -Ss --url 'http://localhost:8080/company'
```

Those requests turned into SQL like
```sql
CALL DOLT_CHECKOUT ('main');
SELECT `company`.`name` AS `name`, `company`.`modified_at` AS `modified_at`, `company`.`id` AS `id`, `company`.`bank_account` AS `bank_account`, `company`.`estimated_size` AS `estimated_size` FROM `company` ORDER BY `company`.`name` ASC LIMIT 0, 20
```

(You can uncomment JDBC logging in `application.yaml`)

Now we can see 20-23% CPU of dolt in `top`.
![Screenshot from 2023-01-20 18-44-58](https://user-images.githubusercontent.com/3160384/213728006-ef6e2c5a-ff60-4af4-89da-2db672515cc5.png) - we have 3 every-second workers

If we open 10 every-second workers we will have
70-100% cpu
![Screenshot from 2023-01-20 19-07-37](https://user-images.githubusercontent.com/3160384/213732105-24627390-f54f-428d-8695-83a3dfe4bcf8.png)

Note that it isn't special load tool like `wrk`, it's just every-second requests produced by `watch`.

I tried to create indexes
```sql
ALTER TABLE company
    ADD INDEX company_name USING HASH (name);

ALTER TABLE company
    ADD INDEX company_name USING BTREE (name);
```
but it doesn't change the CPU load.

# Problem 2 - high CPU and linear time of many branch creation

Here I remained working web app and stopped all watch-based workers.
If we start creating branches again (watch'es still makes requests) CPU of dolt will be 120%
```
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
```
![Screenshot from 2023-01-20 18-54-24](https://user-images.githubusercontent.com/3160384/213728811-d90d930c-5ae9-46d2-92dc-d88f92184f03.png)

Update: I noticed (with help seven_thousand_companies_by_http) that adding more commits doesn't degrade as fast as adding more branches

Update 2: In other words:
We have 7000 companies, created by `seven_thousand_companies`.
The very first pass of `create_draft_for_each_company` takes Measured time PT6M47.349344168S
The next pass - Measured time PT14M32.157333959S
The next pass - Measured time PT21M55.941600155S
The next pass - Measured time PT28M17.140720097S
I suspect that the culprit is - array of branches, which bumped into filesize limit [here](https://github.com/dolthub/dolt/issues/5135).

**Comment by nkonev** [comment]
Because of sort by non-very good selective company name the results were not good.

**Comment by timsehn** [comment]
@Hydrocharged might as well dig into this one as well since he fixed your other, lots of branches problem.

**Comment by nkonev** [comment]
To make issue more shaped:
When I have many branches, for example, ~21000
```
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#seven_thousand_companies" test

./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
```

Creating a new ones by 
```
./mvnw -DenableLoadTests=true -Dtest="name.nkonev.examples.companies.VolumeTest#create_draft_for_each_company" test
```

consumes about of 100% cpu

**Comment by Hydrocharged** [comment]
@nkonev I apologize for the delay, I initially had some issues working with Docker on Windows, but I've finally got some answers! There are two core issues here. The first being that your `SELECT` statements are consuming a lot of resources, and the second being that you experience increasing slowdown as you add more branches.

For the first issue (slow `SELECT` statements), there are two points of slowdown. The first is that we are doing a full table scan with those queries. Even when you add an index on `company.name`, we don't yet support using indexes on `ORDER BY`. Right now we only support indexes on filters, such as with the `WHERE` clause and on joins. I'll file an issue for us to add this functionality. The second point of slowdown is due to accumulated files as the queries are run. To fix this, you can run garbage collection. We have two modes of garbage collection: **standard** and **shallow**. If you run `dolt gc` from a terminal window, then by default it runs the standard collection, which will give the speed increase. The stored procedure `DOLT_GC()` performs a shallow collection, and in my testing it did give a performance increase, but only about half of the standard option, so try and perform a `dolt gc` from a terminal when you can (requires temporarily stopping the server). Automatic GC is on our roadmap.

For the second issue (slow branching), this deals with an implementation detail for the `dolt_branch_control` table. You can read a bit on how they work here: https://docs.dolthub.com/sql-reference/server/branch-permissions. In summary, the aforementioned table controls which users may modify which branches on which databases. Whenever a user creates a branch, a new entry is added to the table, granting that user `admin` privileges over that branch. In our case, as we're creating tens of thousands of branches, the table becomes increasingly large as we add new `admin` entries for the user, and therefore each query iterates over even more rows.

One feature of the table is that we do not add new entries if an existing entry would already grant the same benefits. For example, if you have an `admin` entry with the branch portion being `'a%'`, then we won't add any more entries for that user when a branch starts with the letter `a`. As the branches are being created by a user named `root` in your example, I ran the following query: `INSERT INTO dolt_branch_control VALUES ('%', '%', 'root', '%', 'admin');`. This resulted in:
```
$ dolt sql -q "SELECT * FROM dolt_branch_control;"
+----------+--------+------+------+-------------+
| database | branch | user | host | permissions |
+----------+--------+------+------+-------------+
| %        | %      | %    | %    | write       |
| %        | %      | root | %    | admin       |
+----------+--------+------+------+-------------+
```
With this change, the cost of adding new branches dropped significantly, with adding branches 21,000->28,000 spending a similar amount of time as adding branches 0->7,000, which essentially fixes the issue completely (at least for this test case).

I'm also going to investigate changing how the `dolt_branch_control` tables work. We do an array search to support pattern matching, but we could also do a tree search. For smaller branch counts (up to 100), my initial testing showed that the array search was faster (and easier to implement/maintain), but I'd expect a tree search to outperform the array search when dealing with counts over 10,000. If you're able to add that entry to the `dolt_branch_control` table then that would be great, otherwise I can assign a higher priority to this investigation (such as if you'll have many unknown users each creating their own branches, rather than a single user or a small set of known users).

**Comment by Hydrocharged** [comment]
Also, I'll mention that the increasing file size of `branch_control.db` doesn't have much of an impact at all. I modified the dolt binary (based on `v0.52.8` to keep things comparable) to do two entry comparisons and return true, similar to what it would do with the above fix, and the times were identical.

**Comment by nkonev** [comment]
About the second problem:
>  If you're able to add that entry to the dolt_branch_control table

@Hydrocharged  Thank you a lot for response. Yes, I'm able to apply your SQL. I've applied it, so I have
```
SELECT * FROM dolt_branch_control;

%,%,%,%,write
%,%,root,%,admin
```

I have the following results of `create_draft_for_each_company`
```
Measured time PT4M41.325401001S
Measured time PT7M24.038263521S
Measured time PT10M14.348827364S
```

The speed was increased, but there is linear degradation. May be I'm doing something wrong. I performed insert into `dolt_branch_control` in migrations, they implicitly wraps all statements in transaction. I will move insert in other place. 
I didn't perform any GC. 
Did you perform GC between `create_draft_for_each_company` test runs ?

**Comment by Hydrocharged** [comment]
Yes, you'll need to perform GC between the `create_draft_for_each_company` runs, and also verify that there are only two entries in `dolt_branch_control` via `SELECT count(1) FROM dolt_branch_control;` after the first run to make sure we're not creating new branch entries. If we still are, then we should change the user from `root` to the name of the user that is in those entries.

**Comment by nkonev** [comment]
Thank you, @Hydrocharged 

With both `INSERT INTO dolt_branch_control VALUES ('%', '%', 'root', '%', 'admin');` and `dolt gc` before every `create_draft_for_each_company` test I have
Measured time PT3M34.363319853S (0->7000)
Measured time PT4M30.713584004S (7000->14000)
Measured time PT5M11.280711257S (14000->21000)
Measured time PT6M37.215517519S (21000->28000)

It's much better result.

I'm looking for both 
a) moving our table search to a tree search
b) [Automatic](https://github.com/dolthub/dolt/pull/4959) GC 
as well as using [indexes](https://github.com/dolthub/go-mysql-server/issues/1551) in `ORDER BY`

**Comment by Hydrocharged** [comment]
Fantastic to hear! I'll close this issue for now, as I've created another issue for `ORDER BY`, and it looks like everything has been resolved.

**Comment by bpf120** [comment]
Hi @nkonev, Thanks for filing this issue. We'd love to learn more about your Dolt use case. It really helps us out. Feel free to come by [Discord](https://discord.com/invite/RFwfYpu) or send me an [email](mailto:brianf@dolthub.com) if you want to share.

**Comment by nkonev** [comment]
@bpf120 Hi,
check out here https://github.com/dolthub/dolt/issues/5135#issue-1530809764 - there I wrote my use case

**Comment by Hydrocharged** [comment]
@nkonev Just wanted to let you know that, as of https://github.com/dolthub/dolt/pull/5238, creating branches is _significantly_ faster! These changes are in the latest PR.

**Comment by nkonev** [comment]
@Hydrocharged Thank you! I saw your PR. Will check without inserting in the branch control table.

**Comment by Hydrocharged** [comment]
Sounds good! I do want to mention that it will not outperform inserting that additional row, as that's the intended and designed use of the table. For those cases where users and branches are not known ahead of time and will be quasi-random, then this is a significant speed up.

**Comment by nkonev** [comment]
Test with `dolthub/dolt-sql-server:0.52.16`, results are:
```
Measured time PT2M18.115461131S (0-7000 branches)
- call dolt_gc() -
Measured time PT2M51.115057602S (7001-14000)
- call dolt_gc() -
Measured time PT3M47.193516366S (14001-21000)
- call dolt_gc() -
Measured time PT3M49.033888737S (21001-28000)
```

**Comment by Hydrocharged** [comment]
Results are far better than before! Wonderful
